### PR TITLE
Refactor refreshExtension util function

### LINKED ISSUE #138: Refactor refreshExtensions saga
`refreshExtensions` saga is run after every extension related transaction and its responsibility in the Dapp was to refetch extension queries so that the UI shows the most up-to-date data. 

In the CDapp there's a race condition between the block-ingestor picking up an extension event and writing data to the db and the `refreshExtensions` saga trying to get the updated query results, usually lost by the saga. 

I discussed this with Raul and I think the best solution would be to use "optimistic UI" approach, that is manually update Apollo cache with the expected transaction result (e.g. `ExtensionInstalled` should result in a new `ColonyExtension` entry, `ExtensionDeprecated` should set the `isDeprecated` field to `true` or `false`). That way, the UI gets updated nearly instantly, and in the worst case scenario, any wrong data will be corrected after the next page refresh (as long as the db has been updated).

There's a couple of ways of achieving that in Apollo - https://www.apollographql.com/docs/react/caching/cache-interaction/. The saga itself will have to be modified to accept the payload related to a particular event and there's gonna have to be a switch for handling all the different extension events.

Should be branched off #121

### PR DESCRIPTION
**Pull request by jakubcolony**
## Description

As described in [the issue](https://github.com/JoinColony/colonyCDapp/issues/138), this PR applies the optimistic UI pattern to refreshing extension after a transaction has succeeded.

There are event-specific util functions that modify the Apollo cache with predicted results of the event, which results in an instant update to the UI and avoids the race condition with the block-ingestor. 

It's possible this will lead to inconsistencies in edge cases (try installing an extension and then quickly going back to the list of extensions - it will still show as available) but most of time it provides a much better UX while keeping the solution relatively simple (we can avoid polling/using subscriptions).

There are two issues to follow up:
 - Porting refreshUpgradedExtension - #151 
 - Porting refreshEnabledExtension - #152 
Both are blocked, hence not included in this PR.

## Testing
Install, uninstall, deprecate, re-enable the extensions and the UI should be updated immediately.

Resolves #138

### CODE REVIEW

**Review by willm30 — COMMENTED** [review]
I am having trouble with this one.

I've just come from your other pr so don't know if that's got anything to do with it.

However, when I click install I see:

noinst

With no ui feedback that it's installed. But now when I refresh, it still doesn't show as installed, even though it is (which leads to the following when clicking install):

alredyins

**Comment by Julianahlk** [comment]
@jakubcolony I am also getting the error, "Missing field 'currentVersion' while writing result". Maybe this has do with the recent update `currentVersion: version` in Extension fragment?

**Comment by jakubcolony** [comment]
@willm30 @Julianahlk Apologies for the broken stuff, apart from the obvious rebase conflicts that I missed there was also a more concealed error with the ingestor docker image which only became apparent if you ran it twice or more and resulted in an old version of block-ingestor being run (hence the extension wasn't showing as installed even after you refreshed @willm30)

All should be fixed now

**Comment by jakubcolony** [comment]
@Julianahlk that's a bit odd, it just works on my side. Do you see anything in the console?

File: src/redux/sagas/utils/refreshExtension.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  };
+
+  saveExtensionInCache(modifiedExtension);
+};
+
+export const refreshUninstalledExtension = (
+  colonyAddress: string,
+  extensionId: string,
+) => {
+  const apolloClient = getContext(ContextModule.ApolloClient);
```
[the remark this answers is not in the document]
Comment (jakubcolony) [reply src/redux/sagas/utils/refreshExtension.ts]: Could definitely do that! 🙂

**Comment by Julianahlk** [comment]
> @Julianahlk that's a bit odd, it just works on my side. Do you see anything in the console?

@jakubcolony I restarted my computer tried again and everything is working as expected now. 👍

**Review by Julianahlk — APPROVED** [review]
Working well for me now. 👍

**Review by willm30 — APPROVED** [review]
Working for me! 

Just one question. In the issue, you mention that:

> in the worst case scenario, any wrong data will be corrected after the next page refresh

Is this likely to happen here? Presumably we know exactly what the block ingestor will write to the db and are writing the same info to the apollo cache. Are there any circumstances in which these would deviate from one another?

**Comment by jakubcolony** [comment]
>Is this likely to happen here? Presumably we know exactly what the block ingestor will write to the db and are writing the same info to the apollo cache. Are there any circumstances in which these would deviate from one another?

It is rather unlikely, and, if there's ever any discrepancy between apollo cache and the data block ingestor writes, it will be corrected after the page refresh. The biggest issue I could think of is the `installedAt` timestamp might be a few second (miliseconds?) off compared to what block ingestor will save. So not a big deal at all.

**Pull request merged**